Framework/Kernel/inc/MantidKernel/NearestNeighbours.h

```diff
@@ -0,0 +1,196 @@
+#ifndef MANTID_KERNEL_NEARESTNEIGHBOURS_H_
+#define MANTID_KERNEL_NEARESTNEIGHBOURS_H_
+
+#include "MantidKernel/DllConfig.h"
+#include "MantidKernel/ANN/ANN.h"
+#include "MantidKernel/make_unique.h"
+
+#include <Eigen/Core>
+#include <vector>
+
+/**
+  NearestNeighbours is a thin wrapper class around the ANN library for finding
+  the k nearest neighbours.
+
+  Given a vector of Eigen::Vectors this class will generate a KDTree. The tree
+  can then be interrogated to find the closest k neighbours to a given position.
+
+  This classes is templated with a parameter N which defines the dimensionality
+  of the vector type used. i.e. if N = 3 then Eigen::Vector3d is used.
+
+  @author Samuel Jackson
+  @date 2017
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+
+namespace Mantid {
+namespace Kernel {
+
+//------------------------------------------------------------------------------
+// Helper classes
+//------------------------------------------------------------------------------
+
+/**
+ * NNDataPoints is a thin RAII wrapper class around the ANNpointArray type. This
+ * takes care of the proper allocation and deallocation of memory.
+ */
+class NNDataPoints {
+public:
+  /** Construct a new set of data points
+   *
+   * @param nPts :: the number of data points
+   * @param nElems :: the number of elements for each point
+   */
+  NNDataPoints(const size_t nPts, const size_t nElems) : m_nPts(nPts) {
```

Comment:
Could we use `int` in the signature as the `size_t` type implies we aren't casting down to int.